### ISSUE TITLE
no 6; broken

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9933]
Migrated from [rt.perl.org#70075](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=70075) (status was 'resolved')

Searchable as RT70075$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



I remember that Rafael put a fair amount of work in to make 'no 6' work\.

$ \./perl \-Ilib \-e 'no 6;'
Feature bundle "6\.0\.0" is not supported by Perl 5\.11\.1 at \-e line 1
BEGIN failed\-\-compilation aborted at \-e line 1\.
BEGIN failed\-\-compilation aborted at \-e line 1\.

Well\, it's broken now\.

That \*should\* work without error\, as we're running on a sufficiently old
version\. Contrast with&#8203;:

$ \./perl \-Ilib \-e 'no v5\.11;'
Perls since v5\.11\.0 too modern\-\-this is v5\.11\.1\, stopped at \-e line 1\.
BEGIN failed\-\-compilation aborted at \-e line 1\.

Which is the correct behaviour\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.11.1:

Configured by nclark at Mon Oct 26 11:01:14 GMT 2009.

Summary of my perl5 (revision 5 version 11 subversion 1) configuration:
  Commit id: 0abd0d78a73da1c4d13b1c700526b7e5d03b32d4
  Platform:
    osname=linux, osvers=2.6.24-24-generic, archname=i686-linux
    uname='linux vun---ha-ha-ha.cambridge.netbanx.com 2.6.24-24-generic #1 smp fri sep 18 16:49:39 utc 2009 i686 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache gcc -Dld=gcc -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-g -Uusethreads -Uuse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.11.1-62-g0abd0d7 -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.11.1-62-g0abd0d7 -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache gcc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.2.4 (Ubuntu 4.2.4-1ubuntu4)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.7.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.11.1:
    lib
    /home/nclark/Sandpit/snap5.9.x-v5.11.1-62-g0abd0d7/lib/perl5/site_perl/5.11.1/i686-linux
    /home/nclark/Sandpit/snap5.9.x-v5.11.1-62-g0abd0d7/lib/perl5/site_perl/5.11.1
    /home/nclark/Sandpit/snap5.9.x-v5.11.1-62-g0abd0d7/lib/perl5/5.11.1/i686-linux
    /home/nclark/Sandpit/snap5.9.x-v5.11.1-62-g0abd0d7/lib/perl5/5.11.1
    .


Environment for perl 5.11.1:
    HOME=/home/nclark
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nclark/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
keley@internode.on.net - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

Fixing this simply requires moving the code that loads feature\.pm into  
the preceding block\.

**Comment by p5pRT** [comment]
### From @cpansprout

<details><summary>Inline Patch</summary>

```diff
diff -Nurp blead/pp_ctl.c bleadcopy3/pp_ctl.c
--- blead/pp_ctl.c	2009-11-13 08:54:10.000000000 -0800
+++ bleadcopy3/pp_ctl.c	2009-11-16 20:43:10.000000000 -0800
@@ -3268,23 +3268,30 @@ PP(pp_require)
 			SVfARG(vnormal(PL_patchlevel)));
 		}
 	    }
-	}
 
-        /* We do this only with use, not require. */
-	if (PL_compcv &&
-	  /* If we request a version >= 5.9.5, load feature.pm with the
-	   * feature bundle that corresponds to the required version. */
-		vcmp(sv, sv_2mortal(upg_version(newSVnv(5.009005), FALSE))) >= 0) {
-	    SV *const importsv = vnormal(sv);
-	    *SvPVX_mutable(importsv) = ':';
-	    ENTER_with_name("load_feature");
-	    Perl_load_module(aTHX_ 0, newSVpvs("feature"), NULL, importsv, NULL);
-	    LEAVE_with_name("load_feature");
-	}
-	/* If a version >= 5.11.0 is requested, strictures are on by default! */
-	if (PL_compcv &&
-		vcmp(sv, sv_2mortal(upg_version(newSVnv(5.011000), FALSE))) >= 0) {
-	    PL_hints |= (HINT_STRICT_REFS | HINT_STRICT_SUBS | HINT_STRICT_VARS);
+	    /* We do this only with use, not require. */
+	    if (PL_compcv &&
+	     /* If we request a version >= 5.9.5, load feature.pm with the
+	      * feature bundle that corresponds to the required version. */
+		    vcmp(sv, sv_2mortal(upg_version(newSVnv(5.009005),
+			FALSE))) >= 0) {
+		SV *const importsv = vnormal(sv);
+		*SvPVX_mutable(importsv) = ':';
+		ENTER_with_name("load_feature");
+		Perl_load_module(
+		    aTHX_ 0, newSVpvs("feature"), NULL, importsv, NULL
+		);
+		LEAVE_with_name("load_feature");
+	    }
+	    /* If a version >= 5.11.0 is requested, strictures are on
+	     * by default! */
+	    if (PL_compcv &&
+		    vcmp(sv, sv_2mortal(upg_version(newSVnv(5.011000),
+			FALSE))) >= 0) {
+		PL_hints |= (
+		    HINT_STRICT_REFS | HINT_STRICT_SUBS | HINT_STRICT_VARS
+		);
+	    }
 	}
 
 	RETPUSHYES;
diff -Nurp blead/t/lib/feature/implicit bleadcopy3/t/lib/feature/implicit
--- blead/t/lib/feature/implicit	2009-09-07 04:40:06.000000000 -0700
+++ bleadcopy3/t/lib/feature/implicit	2009-11-16 20:42:58.000000000 -0800
@@ -60,3 +60,7 @@ BEGIN {
 }
 EXPECT
 Helloworld
+########
+# no implicit features with 'no'
+eval "no " . ($]+1); print $@;
+EXPECT
```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun\, Mar 14\, 2010 at 02&#8203;:06&#8203;:22PM \-0700\, Father Chrysostomos wrote&#8203;:
> Fixing this simply requires moving the code that loads feature\.pm into  
> the preceding block\.

This was fixed in faee19b5\, but no test was included\, so I've applied
the test part of the patch\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
